### ISSUE TITLE
Apply garp_delay to unsolicited ARP when lower prio advert received

### ISSUE THREAD

**Issue by flicker581** [acassen/keepalived#277]
We have garp_delay to send ARP twice when becoming master. But when the master receives lower prio advert, it sends ARP only once. This happens on a congested network. Then, with high probability, hosts on this network have wrong ARP entries, because there is a race between ARP from two routers.

Of course, garp_refresh is a working workaround.

I propose to add garp_delay logic to the case of "lower prio advert received".

Sorry for this not being pull request, but, anyway, I am not sure that this is the right way to fix the issue. But it works for me.

    diff -ru keepalived-1.2.19-p/keepalived/include/vrrp_scheduler.h keepalived-1.2.19-pp/keepalived/include/vrrp_scheduler.h
    --- keepalived-1.2.19-p/keepalived/include/vrrp_scheduler.h     2015-07-07 15:39:19.000000000 +0300
    +++ keepalived-1.2.19-pp/keepalived/include/vrrp_scheduler.h    2016-03-14 12:06:19.000000000 +0300
    @@ -62,4 +62,6 @@
    extern int vrrp_dispatcher_init(thread_t *);
    extern int vrrp_read_dispatcher_thread(thread_t *);
     
    +extern int vrrp_gratuitous_arp_thread(thread_t * thread);
    +
    #endif
    diff -ru keepalived-1.2.19-p/keepalived/vrrp/vrrp.c keepalived-1.2.19-pp/keepalived/vrrp/vrrp.c
    --- keepalived-1.2.19-p/keepalived/vrrp/vrrp.c  2015-07-07 16:06:11.000000000 +0300
    +++ keepalived-1.2.19-pp/keepalived/vrrp/vrrp.c 2016-03-14 13:26:01.000000000 +0300
    @@ -1325,6 +1325,10 @@
                    }
                    vrrp_send_adv(vrrp, vrrp->effective_priority);
                    vrrp_send_link_update(vrrp, vrrp->garp_rep);
    +                thread_add_timer(master, vrrp_gratuitous_arp_thread,
    +                                 vrrp,
    +                                 (vrrp->garp_delay) ?
    +                                        vrrp->garp_delay : VRRP_GARP_DELAY);
                    return 0;
            } else if (hd->priority == 0) {
                    vrrp_send_adv(vrrp, vrrp->effective_priority);

**Comment by pqarmitage** [comment]
Your patch looks just about right. The only change is that vrrp->garp_delay == 0 means that no second set of ARP messages is to be sent (set by configuration parameter garp_master_delay).

To increase the likelihood of at least some of the ARP messages not being lost, you could also use garp_master_repeat parameter, which determines how many ARP messages are sent each time the ARP messages are sent (this of course will increase the network load, especially if you have multiple VIPs).

If your network is so heavily congested that 3 consecutive VRRP messages are dropped, then it seems likely that overall behaviour is going to be unreliable anyway.

Commit 4295e87 implements the patch as above, but with the check that vrrp->garp_delay != 0. Commit 88a4b84 allows configuring of the action after a lower priority advert is received separately from the action after transition to master.
